src/read/xz.rs

```diff
@@ -0,0 +1,330 @@
+                ));
+            }
+            if *reader.count % 4 != 0 {
+                let mut b = vec![0u8; 4 - *reader.count % 4];
+                reader.read_exact(&mut b)?;
+            }
+            *reader.count = 0;
+            return self.read(buf);
+        } else {
+            // block
+            let header_size = (b[0] as u16 + 1) * 4 - 1 - 4;
+            let mut b = vec![0u8; header_size as usize];
+            reader.read_exact(b.as_mut_slice())?;
+            digest.update(&b);
+            let mut reader = BufReader::new(b.as_slice());
+            let mut b = [0u8; 1];
+            reader.read_exact(&mut b)?;
+            let flags = b[0];
+            let num_filters = (flags & 0x03) + 1;
+
+            if flags & 0x3C != 0 {
+                return Err(Error::new(
+                    std::io::ErrorKind::InvalidData,
+                    "Invalid XZ block flags",
+                ));
+            }
+            if flags & 0x40 != 0 {
+                get_multibyte(&mut reader, None)?;
+            }
+            if flags & 0x80 != 0 {
+                get_multibyte(&mut reader, None)?;
+            }
+            for _ in 0..num_filters {
+                let filter_id = get_multibyte(&mut reader, None)?;
+                if filter_id != 0x21 {
+                    return Err(Error::new(
+                        std::io::ErrorKind::InvalidData,
+                        "Unsupported XZ filter ID",
+                    ));
+                }
+                let properties_size = get_multibyte(&mut reader, None)?;
+                if properties_size != 1 {
+                    return Err(Error::new(
+                        std::io::ErrorKind::InvalidData,
+                        "Unsupported XZ filter properties size",
+                    ));
+                }
+                reader.read_exact(&mut b)?;
+            }
+            while reader.read(&mut b)? > 0 {
+                if b[0] != 0 {
+                    return Err(Error::new(
+                        std::io::ErrorKind::InvalidData,
+                        format!("Invalid XZ block padding {}", b[0]),
+                    ));
+                }
+            }
+        }
+        let mut b = [0u8; 4];
+        reader.read_exact(&mut b)?;
+        if digest.finalize().to_le_bytes() != b {
+            return Err(Error::new(
+                std::io::ErrorKind::InvalidData,
+                "Invalid XZ block header CRC32",
+            ));
+        }
+        let mut written = 0;
+        let mut total = 0;
+        Lzma2Decoder::new().decompress(
+            &mut reader,
+            &mut BufWriter {
+                inner: buf,
+                written: &mut written,
+                rest: &mut self.buf,
+                total: &mut total,
+            },
+        )?;
+
+        let unpadded_size = *reader.count - block_begin;
+        self.records.push((unpadded_size, total));
+
+        if unpadded_size % 4 != 0 {
```

Comment:
Combine this and the following read into one, by summing the total length in a mutable variable.